patches/minecraft/net/minecraft/world/entity/SpawnPlacements.java.patch

```diff
@@ -26,5 +34,20 @@
 +          if (predicate == null) return net.minecraft.world.level.NaturalSpawner.canSpawnAtBody(this, world, pos, type);
 +          return predicate.test(world, pos, type);
 +      }
++   }
++
++   // ******* FORGE START. INTERNAL USE ONLY! ****** //
++   public static void fireSpawnPlacementEvent()
++   {
++       Map<EntityType<?>, net.minecraftforge.event.entity.SpawnPlacementRegisterEvent.MergedSpawnPredicate> map = Maps.newHashMap();
++       f_21750_.forEach((type, data) -> {
++           map.put(type, new net.minecraftforge.event.entity.SpawnPlacementRegisterEvent.MergedSpawnPredicate(data.f_21769_, data.f_21768_, data.f_21767_));
++       });
++       net.minecraftforge.common.ForgeHooks.onSpawnPlacementRegister(map);
++   }
++
++   public static void registerAll(Map<EntityType<?>, net.minecraftforge.event.entity.SpawnPlacementRegisterEvent.MergedSpawnPredicate> map)
++   {
++       map.forEach(((entityType, merged) -> f_21750_.put(entityType, new SpawnPlacements.Data(merged.getHeightmapType(), merged.getSpawnType(), merged.build()))));
```

Comment:
Does not need to be such a large patch. Can be simplified with a call to the hooks class:

```java
public static void modifyPlacements() {
    Map<EntityType, MergedSpawnPred> map = ForgeHooks.buildPlacements(DATA_BY_TYPE);
    map.forEach(DATA_BY_TYPE.put());
}
```

Reply:
yes and no, MergedSP and SP.Data are not convertible between each other, and SP.Data is private. So I cannot read anything about the SP.Data in ForgeHooks.

Definitely can be made smaller though